linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala

```diff
@@ -849,8 +861,13 @@ private final class Analyzer(config: CommonPhaseConfig,
       implicit val from = FromExports
 
       tryLookupStaticLikeMethod(MemberNamespace.StaticConstructor,
-          StaticInitializerName).foreach {
-        _.reachStatic()(fromAnalyzer)
+          StaticInitializerName).foreach { initializer =>
+        initializer.reachStatic()(fromAnalyzer)
+        staticInitializerModule =
+          moduleSelectors.classifyClass(className.nameString)
+        if (staticInitializerModule.isEmpty) {
+          _errors += ???
```

Comment:
This is problematic for the following:

```scala
object Foo {
  @JSExportTopLevel("myField")
  val x = 1
}

val moduleSpec = ModuleSpec(modName).withSelector(ModuleSpec.Selector.exportName("myField"))
```

This will trigger an error here, because `Foo`'s static initializer (created for the static field) is not classified into a leaf module. In this specific example, this is not a problem, because static initializers are always emitted with their classes and the top level export depends on it.

However, when static initializers are used for reflection, they need to be referenced from some leaf module. Otherwise they never register themselves.

This is a specification problem for module splitting in the presence of static initializers. The easiest fix is to simply disallow classifying exports differently based on their name (but only based on class name). But that seems somewhat arbitrary.

@sjrd WDYT?

Reply:
This is much more problematic than that. From first principles, our spec today for static initializers and they are (all) executed when "the module" is loaded. In theory, that would mean that they should all be executed at startup, whatever leaf module is used, and that consequently they become dependencies of *all leaf modules*. (They can be stored in their class' module, that's alright, as long as all leaf modules make sure to depend on them.)

Those semantics are IMO the only correct semantic for reflection, since we cannot know who will want what when. Any other semantics risks creating non-deterministic results for `Reflect` uses. (An alternative, semantically correct as well, would be that the module that holds `Reflect` itself depends on all those static initializers...)

For the module initializer of exports, however, that would kill the separation of top-level exports in several modules, as far as transitive dependencies are concerned (particularly for the "biggest modules" splitting strategy). For those we clearly only want them depended on by their exports.

Perhaps we'll need to split the semantics of our static initializers :s Or rethink how they work entirely.

I'm going to keep thinking about this.

Reply:
I've been thinking about this as well and have been wondering: What if we put the burden of deciding which module something is in on the compiler?

Consequences:

- We'd need a "default module" for backwards compatibility.
- For top-level exports, that would just be another annotation parameter.
- For static initializers, the compiler can make a more informed decision, as it has semantic information about the use of static initializers. (to solve the `Reflect` use case properly, we'd need some form of language to express the modules that need to reach the static initializers... not sure if that works).

Reply:
That could work.

For exports, is there ever a valid use case for something different than where the enclosing class goes? The exports themselves and the static initializer should be in the same module, and the static initializer depends on the class' module accessor. So they'll always end up together anyway.

Reply:
For `Reflect`, to me the natural, semantically correct choice is to go in the same module as `Reflect` itself. Because you need them to run if and only if you use one of the methods of `Reflect`.

Reply:
> For exports, is there ever a valid use case for something different than where the enclosing class goes?

No, I don't think so.

> The exports themselves and the static initializer should be in the same module.

This is no correct.  Splitting exports away from the classes they are defined in is necessary, since other modules might also depend on the class, so the class itself needs to be in an "internal" (i.e. non-leaf) module.

As a consequence, the current scheme allows exports of the same class to go into different leaf modules, because it's "free".

> For Reflect, to me the natural, semantically correct choice is to go in the same module as Reflect itself.

Yes. The question there is: Do we really want to allow a spec of the form: "Wherever X is"? If the answer is yes, it's unclear to me how we would solve backwards compatibility here :-/

On a related note: Your wording suggests that we put the static initializers themselves in the relevant modules. The current design always leaves them with the class, it just force-imports the class form things that "reach" the static initializer. This ensures that a class can never be used without the static initializer running. I think we probably want to keep that. This also avoids the problem of having multiple static initializers (they just might be force-imported form multiple spots).

Reply:
> The current design always leaves them with the class, it just force-imports the class form things that "reach" the static initializer. This ensures that a class can never be used without the static initializer running. I think we probably want to keep that.

Yes, absolutely. I wasn't very clear indeed. The current design is good.

> The question there is: Do we really want to allow a spec of the form: "Wherever X is"? If the answer is yes, it's unclear to me how we would solve backwards compatibility here :-/

Our existing static initializers follow a simple pattern. We can have a simple, if gross, hack to transition them.

Reply:
So that would basically amount to introducing a "reverse dependency". Where a class arbitrarily could specify that another one "depends" on it.

I'm wondering if we should introduce "registration" as a paradigm. So instead of baking this into the static initializer, we could have a "registerTo" thingy. This would, in the future, allow us to even eliminate these dependencies if they registry isn't reachable.

Reply:
It's difficult to imagine what a registration feature should look like without having at least a second use case. But that seems like an idea to explore.

Reply:
Java's SPI comes to mind: https://docs.oracle.com/javase/tutorial/ext/basics/spi.html (also see https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html)